### DISCUSSION TITLE
Style changing commands

### DISCUSSION THREAD

**Discussion by canedha** [pymupdf/PyMuPDF#1520] [Looking for help] [answered]
Hi @JorjMcKie,
hopefully quick question for you (or anybody else who knows):
The style of an object (line, rectangle etc) is set e. g. by operator d for stroke and then drawn by s. If you directly draw another line it uses the same style. Are there any operators e. g. q/Q that resets/changes the objects style setting?
Hope you are doing well and a happy new year!

**Comment by JorjMcKie** [comment, marked as answer]
Happy New Year to you too!

You are right: `"q"` and `"Q"` are defined as the stacking commands for the "graphics state". Any graphics setting made within their scope is reset to the corresponding value preceeding the `"q"`.
Which PDF commands these are (it's not all of them) is best looked up in the [PDF manual](https://www.adobe.com/content/dam/acom/en/devnet/pdf/pdfs/PDF32000_2008.pdf), page 125+.

In PyMuPDF, class `Shape` controls this: The `.finish()` method takes care of this wrapping. Any draw command preceeding it will be put under the effect of the parameters given to `finish()`.
If you instead use the same-named `Page` methods, like `page.draw_rect()`, then the corresponding shape method is invoked - **_plus_** using the extra parameters, a `finish()` and a `commit()` will be executed. So in this case, the single draw command will be wrapped be q/Q ... and also land in a separate, own `/Contents` object (every `.commit()` will create a new `/Contents` object).

**Reply by canedha** [reply]
@JorjMcKie awesome as always! thx!
